### ISSUE TITLE
[hwmonitor@sylfurd]: housekeeping, TODOs & future changes

### ISSUE THREAD

**Issue by jacobwills** [linuxmint/cinnamon-spices-applets#3011]
**Notify author of applet**
@Hultan 

**Feature Request**
I've compiled a list of some action items that I'd like for us to work on to ensure that this applet project is moving in the best possible direction. See some of the items below and give me your thoughts and feedback. Please let me know what you think or if I should be proposing these ideas in a different format consistent with the contributing guidelines (I didn't see anything directly related to this sort of thing). When these issues begin to actually be worked on, an issue can be created to which an assignee could be designated. We could possibly create a `TODO.md` file to maintain a list of future changes we want to track that would look something like the following outline.

Cheers

---

TODO
====

Housekeeping
--------------------------

- Code formatting/linting overview (make sure indentation, spacing between function names, equality symbols, etc. in assignments are at least consistent throughout to make it easier to read)
- Update class and function declarations to be consistent with the current recommended standard implementation suggested by Linux Mint's documentation. (i.e.`class GraphicalHWMonitorApplet extends Applet.IconApplet {constructor(x,y,z)}` vs using the older Prototype model)
- Copyright claim in _applet.js_ should be removed completely? OR
- Fix the conflicting licensing statements in our code. (GPL and MIT listed in different places)
- Naming of the product should be fixed. Use of _HWMonitor_ or _Foobar_ [probably copy-paste error] or _GHW_ or _Graphical Hardware Monitor_ is presented in different places throughout the code and UI. Correct capitalization.
- Move additional screenshots to a separate folder for better organization and update the image references in the README.md respectively. (obviously leave the _screenshot.png_ in the correct location for it to work with the applet info [page](https://cinnamon-spices.linuxmint.com/applets/view/12) properly)
- Separate change log in to CHANGELOG.md and add updating guidelines (revision numbering rules, etc.)
- Remove needless comments (obvious description of functions, changelog info that's written in code comments, etc.) These should be in the README.md, CHANGELOG.md or somewhere else, if it needs to be stated at all.
- Add useful comments to bits of code that need explanation or change the names of functions and variables to better describe their purpose or usage.
- Draft a Commit/PR Guidelines statement to put in the README.md or CONTRIBUTING.md (link to Cinnamon Dev guidelines, plus our own rules such as: one major change per commit/PR, squash commits of a PR that are very minor like removing comments, debug lines, etc. that should have been removed before committing them anyway, etc.)
- For contributors - add links to GitHub profiles and/or their commit history in this project to see their changes
- Refine complex and hard to follow pieces of code to make it easier to understand/maintain?
- Code optimizations. Research methods to speed up the app if possible. (JavaScript [Cjs], GNOME modules, etc.)?
- Find a way to remove the "crash causing function calls" warning upon user installation.
- Find a way to take advantage of the inclusion of different code for different Cinnamon versions (backport features/bugs to older version? Create updated version with compatibility issues fixed if needed).

Features
--------

Looking at some other applet's implementations to get ideas...
- SpicesUpdate - gives me the idea to use table to add disks that should be monitored and use that to create as many graphs as the user wants.
- Check if dependencies are installed and help the user fetch/install them if possible
- Display colorized warnings if settings are incompatible (related to disk disappearing bug below)
- Add tool-tips to show user warnings, errors, info, suggestions. place on individual graphs if that's possible
- Consolidate tabs into: _General, Processor, Memory, Network, Storage, Battery_
- Improve battery monitoring functionality by adding info from `upower` OR `apcupsd`/`apcaccess` which gives desktop/server users information about UPS batteries connected via USB over the network.
- Improve settings UI by implementing custom widgets OR by writing a custom GTK application from scratch that has more features and handles input from user, provides feedback/interactivity with the panel applet visually, and handling output to config files or logs, etc. even better that the Cinnamon xlet API and with more freedom.
- Add more themes or the ability to theme panels individually AND
- Add the ability to save themes. Also refine the themes implementation to make it easier to work with, both for the user in the GUI and in code for maintaining and recalling saved themes.

Bugs
-----------

- Fail gracefully and/or give warning/error message to the user about selected device being unavailable (happens when the disk disappears from the system like when failed or unmounted [usb drive, hotplug SATA/SAS/NVMe removed, etc.])
- Fix warning "Invalid or null source id used when attempting to run Mainloop.source_remove()"

**Comment by Hultan** [comment]
I think that the todo.md is a great idea, so that we don't have to keep an issue open for stuff like this. I have historically not created issues first, I have just been doing them when I had time to do it, since no one else seemed to be doing them :-)

But if more people (read: you) plan to get involved, I am all for getting a little bit more organized.

**Comment by Hultan** [comment]
@jacobwills I have (at least temporarily) left Cinnamon (it have crashed basically every single day for a year now for me), so feel free to "take ownership" of this applet. Just change the **author** field in **info.json**...I don't think I will be working on this applet for a while...

**Comment by jacobwills** [comment]
@Hultan Thanks for the heads up.
